src/persist-types/src/arrow.rs

```diff
@@ -417,3 +423,115 @@ impl<'a> PartialEq for ArrayIdx<'a> {
 }
 
 impl<'a> Eq for ArrayIdx<'a> {}
+
+/// An array with precisely one entry, for use as a lower bound.
+#[derive(Debug, Clone)]
+pub struct ArrayBound {
+    raw: ArrayRef,
+    ord: ArrayOrd,
+    index: usize,
+}
+
+impl ArrayBound {
+    /// Create a new `ArrayBound` for this array, with the bound at the provided index.
+    pub fn new(array: ArrayRef, index: usize) -> Self {
+        Self {
+            ord: ArrayOrd::new(array.as_ref()),
+            raw: array,
+            index,
+        }
+    }
+
+    /// Get the value of the bound.
+    pub fn get(&self) -> ArrayIdx {
+        self.ord.at(self.index)
+    }
+
+    /// Convert to an array-data proto, respecting a maximum data size. The resulting proto will
+    /// decode to a single-row array, such that `ArrayBound::new(decoded, 0).get() <= self.get()`,
+    /// which makes it suitable as a lower bound.
+    pub fn to_proto_lower(&self, max_len: usize) -> Option<ProtoArrayData> {
+        let array_data = match self.raw.len() {
+            0 => return None,
+            1 => self.raw.to_data(),
```

Comment:
Does this do the right thing if the input is a slice of some larger array? I.e. should we just always fall into the case below?

Reply:
Yeah, I think that's best. (The answer is some flavour of "not always", and it seems enough that it's worth the very marginal cost to have a simpler one.)

Reply:
Done! Definitely feels more straightforward.